lib/src/piece/case.dart

```diff
@@ -89,22 +81,18 @@ class CaseExpressionPiece extends Piece {
     // If there is a split guard, then indent the pattern past it.
     var indentPatternForGuard = !_canBlockSplitPattern &&
         !_patternIsLogicalOr &&
-        (state == _beforeWhen || state == _beforeWhenAndBody);
+        state == _beforeWhenAndBody;
 
     if (indentPatternForGuard) writer.pushIndent(Indent.expression);
 
-    writer.pushAllowNewlines(allowNewlineInPattern);
-    writer.format(_pattern);
-    writer.popAllowNewlines();
+    writer.format(_pattern, allowNewlines: allowNewlineInPattern);
 
     if (indentPatternForGuard) writer.popIndent();
 
     if (_guard case var guard?) {
       writer.pushIndent(Indent.expression);
-      writer.pushAllowNewlines(allowNewlineInGuard);
-      writer.splitIf(state == _beforeWhen || state == _beforeWhenAndBody);
```

Comment:
...and then the piece writes a newline *itself* when `state` is `_beforeWhen`, invalidating its own state and preventing `_beforeWhen` from ever being a valid solution. Oops.